homeassistant/components/mobile_app/__init__.py

```diff
@@ -0,0 +1,326 @@
+    return (SecretBox.KEY_SIZE, decrypt)
+
+
+def _decrypt_payload(key, ciphertext):
+    """Decrypt encrypted payload."""
+    try:
+        keylen, decrypt = get_cipher()
+    except OSError:
+        _LOGGER.warning(
+            "Ignoring encrypted payload because libsodium not installed")
+        return None
+
+    if key is None:
+        _LOGGER.warning(
+            "Ignoring encrypted payload because no decryption key known")
+        return None
+
+    key = key.encode("utf-8")
+    key = key[:keylen]
+    key = key.ljust(keylen, b'\0')
+
+    try:
+        message = decrypt(ciphertext, key)
+        message = json.loads(message.decode("utf-8"))
+        _LOGGER.debug("Successfully decrypted mobile_app payload")
+        return message
+    except ValueError:
+        _LOGGER.warning("Ignoring encrypted payload because unable to decrypt")
+        return None
+
+
+def context(request):
+    """Generate a context from a request."""
+    user = request.get('hass_user')
+    if user is None:
+        return Context()
+
+    return Context(user_id=user.id)
+
+
+async def handle_webhook(hass: HomeAssistantType, webhook_id: str, request):
+    """Handle webhook callback."""
+    device = device_for_webhook_id(hass, webhook_id)
+
+    req_data = await request.json()
+
+    try:
+        req_data = WEBHOOK_PAYLOAD_SCHEMA(req_data)
+    except vol.Invalid as ex:
+        return json_response(vol.humanize.humanize_error(req_data, ex),
+                             status=HTTP_BAD_REQUEST)
+
+    webhook_type = req_data[ATTR_WEBHOOK_TYPE]
+
+    webhook_payload = req_data.get(ATTR_WEBHOOK_DATA, {})
+
+    if req_data[ATTR_WEBHOOK_ENCRYPTED]:
+        enc_data = req_data[ATTR_WEBHOOK_ENCRYPTED_DATA]
+        webhook_payload = _decrypt_payload(device[CONF_SECRET], enc_data)
+
+    try:
+        data = WEBHOOK_SCHEMAS[webhook_type](webhook_payload)
+    except vol.Invalid as ex:
+        return json_response(vol.humanize.humanize_error(webhook_payload, ex),
+                             status=HTTP_BAD_REQUEST)
+
+    if webhook_type == WEBHOOK_TYPE_CALL_SERVICE:
+
+        with AsyncTrackStates(hass) as changed_states:
+            try:
+                await hass.services.async_call(data[ATTR_DOMAIN],
+                                               data[ATTR_SERVICE],
+                                               data[ATTR_SERVICE_DATA],
+                                               True,
+                                               context(request))
+            except (vol.Invalid, ServiceNotFound):
+                raise HTTPBadRequest()
+
+        return json_response(changed_states)
+    elif webhook_type == WEBHOOK_TYPE_FIRE_EVENT:
+        event_type = data[ATTR_EVENT_TYPE]
+        hass.bus.fire(event_type, data[ATTR_EVENT_DATA], ha.EventOrigin.remote,
+                      context(request))
+        return json_response({"message": "Event {} fired.".format(event_type)})
```

Comment:
I think that we should just return an empty JSON response? Or maybe even no response and just 200 ?

Reply:
I did it this way to exactly match the outputs of the REST API since that's what the iOS app was already built against. Not opposed to removing this one though.

Reply:
let's drop it.